### PR TITLE
`azurerm_role_assignment` - Support scope to be `/providers/Microsoft.Marketplace`

### PR DESCRIPTION
**Pull request by ms-zhenhua**
fix https://github.com/hashicorp/terraform-provider-azurerm/issues/18387
image

### CODE REVIEW

**Comment by ms-zhenhua** [comment]
Note: `TestAccRoleAssignment_marketPlaceScoped` cannot pass on TC because the SP used by TC does not have authorization to create a `Marketplace Admin` role. The admin should first grant the SP with `owner` access according to this [Doc](https://learn.microsoft.com/en-us/marketplace/create-manage-private-azure-marketplace-new#assign-the-marketplace-admin-role-with-access-control-iam), then run this testcase.

**Review by katbyte — CHANGES_REQUESTED** [review]
@ms-zhenhua - could we update the docs & add that link and an example powershell of how to grant permissions to a SP?

**Comment by ms-zhenhua** [comment]
Hi @katbyte, thank you for reviewing. I have updated the document with the related link which contains the information of how to grant permissions to different identities by Azure Portal, Azure PowerShell, Azure Cli and Rest API. Kindly help review it again. Thanks.

File: internal/services/authorization/parse/role_assignment.go [hunk 1]
```diff
@@ -145,7 +136,15 @@ func RoleAssignmentID(input string) (*RoleAssignmentId, error) {
...
-		return nil, fmt.Errorf("could not parse Role Assignment ID %q", input)
+		re := regexp.MustCompile(`^/providers/(Microsoft.[a-zA-Z]+)/providers/Microsoft.Authorization/roleAssignments/([a-fA-F\d]{8}-[a-fA-F\d]{4}-[a-fA-F\d]{4}-[a-fA-F\d]{4}-[a-fA-F\d]{12})$`)
+		matches := re.FindStringSubmatch(input)
+		if len(matches) != 3 {
+			return nil, fmt.Errorf("could not parse Role Assignment ID %q", input)
+		}
+
+		roleAssignmentId.IsRootLevel = true
+		roleAssignmentId.ResourceProvider = matches[1]
+		roleAssignmentId.Name = matches[2]
```
Comment (manicminer) [inline internal/services/authorization/parse/role_assignment.go:147]: Can we avoid parsing with regular expressions and rather perform simple string matching/splitting? It's better that we test new RPs and explicitly support them.

**Review by manicminer — CHANGES_REQUESTED** [review]
@ms-zhenhua I've assigned the `Marketplace Admin` role so you should be able to test this in TC now. I've left another comment below regarding parsing the scope, if you can rework this I'll be happy to re-review. Thanks!

**Comment by ms-zhenhua** [comment]
Hi @manicminer, thank you for reviewing. The testcase needs to assign `Marketplace Admin` role to a new SP. But the existing SP with `Marketplace Admin` role does not have the permission to assign `Marketplace Admin` role to a new SP. To make the testcase runnable on TC, could you grant the test SP with the `Owner` role of private marketplace?
image

File: website/docs/r/role_assignment.html.markdown [hunk 2]
```diff
@@ -132,15 +132,15 @@ The following arguments are supported:
...
 
 * `role_definition_id` - (Optional) The Scoped-ID of the Role Definition. Changing this forces a new resource to be created. Conflicts with `role_definition_name`.
 
 * `role_definition_name` - (Optional) The name of a built-in Role. Changing this forces a new resource to be created. Conflicts with `role_definition_id`.
 
 * `principal_id` - (Required) The ID of the Principal (User, Group or Service Principal) to assign the Role Definition to. Changing this forces a new resource to be created.
 
-~> **NOTE:** The Principal ID is also known as the Object ID (ie not the "Application ID" for applications).
+~> **NOTE:** The Principal ID is also known as the Object ID (ie not the "Application ID" for applications). To assign Azure roles, the Principal mush have `Microsoft.Authorization/roleAssignments/write` permissions. See [documentation](https://learn.microsoft.com/en-us/azure/role-based-access-control/role-assignments-portal) for more information.
```
[the remark this answers is not in the document]
Comment (ms-zhenhua) [reply website/docs/r/role_assignment.html.markdown]: According to the description of roles on Azure Portal, he calling principal with `Marketplace Admin role` role cannot assign `Marketplace Admin role` to other principals, so I updated this comment as:  
image

**Comment by manicminer** [comment]
@ms-zhenhua Ah, I misunderstood, this has now been assigned. Please could you add a comment to the acctest mentioning the required owner role (with docs link)? Thanks!

**Comment by ms-zhenhua** [comment]
Hi @manicminer, thank you for your update. I have added the comments to the acctest and all related testcases have passed on TC. Kindly help review it again.

**Comment by katbyte** [comment]
@ms-zhenhua - upon some internal discussion we've come to the conclusion that the role assignment resource should probably be broken out into separate ones for each resource type for an improved UX much like we have different resource for policy assignments.

as such for this resource we could extract it out into  `azurerm_role_assignment_marketplace` and leave the old resource to be refactored at a later time.

WDYT?

**Comment by ms-zhenhua** [comment]
Hi @katbyte, totally agree. Then I will temporarily close this PR and create a new one for the new resource `azurerm_role_assignment_marketplace`

**Pull request closed without merging**